src/ttl_map.rs

```diff
@@ -0,0 +1,730 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use dashmap::mapref::entry::Entry as DashMapEntry;
+use dashmap::mapref::one::{Ref, RefMut};
+use dashmap::DashMap;
+use slog::{warn, Logger};
+use std::borrow::Borrow;
+use std::hash::Hash;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+use tokio::sync::oneshot::{channel, Receiver, Sender};
+use tokio::time::Instant;
+
+/// A wrapper around the value of an entry in the map.
+/// It contains the value's ttl.
+pub(crate) struct Value<V> {
+    pub value: V,
+    expires_at: Arc<AtomicU64>,
+    #[cfg(test)]
+    clock: TestClock,
+    #[cfg(not(test))]
+    clock: RealClock,
+}
+
+impl<V> Value<V> {
+    fn new(
+        value: V,
+        log: &Logger,
+        ttl: Duration,
+        #[cfg(test)] clock: TestClock,
+        #[cfg(not(test))] clock: RealClock,
+    ) -> Value<V> {
+        let value = Value {
+            value,
+            expires_at: Arc::new(AtomicU64::new(0)),
+            clock,
+        };
+        value.update_expiration(log, ttl);
+        value
+    }
+
+    /// Get the expiration time for this value. The returned value is the
+    /// number of seconds relative to some reference point (e.g UNIX_EPOCH), based
+    /// on the clock being used.
+    fn expiration(&self) -> u64 {
+        self.expires_at.load(Ordering::Relaxed)
+    }
+
+    /// Update the value's expiration time to (now + TTL).
+    fn update_expiration(&self, log: &Logger, ttl: Duration) {
+        match self.clock.compute_expiration(ttl) {
+            Ok(new_expiration_time) => {
+                self.expires_at
+                    .store(new_expiration_time, Ordering::Relaxed);
+            }
+            Err(err) => {
+                warn!(log, "failed to increment key expiration: {}", err)
+            }
+        }
+    }
+}
+
+/// Map contains the hash map implementation.
+struct Map<K, V> {
+    inner: DashMap<K, Value<V>>,
+    log: Logger,
+    ttl: Duration,
+    shutdown_tx: Option<Sender<()>>,
+
+    #[cfg(test)]
+    clock: TestClock,
+    #[cfg(not(test))]
+    clock: RealClock,
+}
+
+impl<K, V> Drop for Map<K, V> {
+    fn drop(&mut self) {
+        if let Some(shutdown_tx) = self.shutdown_tx.take() {
+            shutdown_tx.send(()).ok();
+        }
+    }
+}
+
+/// TTLMap is a key value hash map where entries are associated with a TTL.
+/// When the TTL for an entry elapses, the entry is removed from the map.
+/// The TTL is reset each time the entry is (re)inserted or read via [`TTLMap::get`],
+/// [`TTLMap::get_mut`] functions, or via the [`TTLMap::entry`] interface.
+/// During tests, the internal clock implementation is driven by [`tokio::time`] so
+/// functions like [`tokio::time::pause`] and [`tokio::time::advance`] can be used.
+#[derive(Clone)]
+pub(crate) struct TTLMap<K, V>(Arc<Map<K, V>>);
```

Comment:
Updated to TtlMap (SessionMap would be confusing here since we already use that for something else), and made the type less generic

Reply:
> I'll also mention this at the meeting on Wednesday, but while thinking about this problem, I think this (and a lot our of architecture discussions) is best solved by using an "Entity Component System" or at the very least a entity map.

If I repeat this back, so that I think I understand it:

We should have the functionality to tie data to connection Sessions (which would be the entity), and then each Filter (or System, in ECS terms) could attach arbitrary data (as components) to each Session.

Therefore each Filter wouldn't need to track and prune information related to each Session, it could just use the underlying Session data storage, and not have to worry about it.

🤔 that's a really interesting idea. Looking forward to chatting about it at the upcoming meeting 👍🏻 

- Did I get that right?

Reply:
> Therefore each Filter wouldn't need to track and prune information related to each Session, it could just use the underlying Session data storage, and not have to worry about it.

Yes,  exactly. And I think we could go even further, and move individual UDP packets to an entity system as well. Our current design is very "Array of Structures" (each ctx iterating over each filter), but I think the "Structure of Arrays" that comes from entity component architectures would address the concerns I've mentioned about a better separation between systems. For example, this is what load balancing would look like if load balancing was implemented using `legion`. The world in this case, is a world of `ReadContext` entities. You can see that we only query for the actual data we need in the filter, and another bonus is that this is parallel for free as long as `LoadBalancer` is `Send + Sync`.

```rust
impl Runnable for LoadBalancer {
    fn run(&mut self, world: &mut World, _resources: &mut Resources) {
        let query = <(&mut UpstreamEndpoints, &SocketAddr)>::query();
        
        // or .par_iter_mut().for_each(|_| ...)
        for (upstream_endpoints, source) for query.iter_mut(world) {
            self.endpoint_chooser.choose_endpoints(upstream_endpoints, source);
        }
    }
}
```

Reply:
sorry, only just had a chance to read this...  should have looked before the call today.

have you looked at fd.io VPP at all?  This feels like something VPP-ish where each filter iterates over all available packets, rather than each packet iterating through the list of filters (unless am misunderstanding it).

(and apologies here if am teaching you to suck eggs).  This model gives performance benefits as the first packet to hit a specific filter results in the filter code being loaded into the CPU instruction cache.   Subsequent packets take advantage of that pre-warmed cache...

Reply:
> This feels like something VPP-ish where each filter iterates over all available packets, rather than each packet iterating through the list of filters (unless am misunderstanding it).

I haven't, I'll have a look at it but you're exactly right.

> This model gives performance benefits as the first packet to hit a specific filter results in the filter code being loaded into the CPU instruction cache. Subsequent packets take advantage of that pre-warmed cache...

Yes, this would provide much greater data locality as it would allow all `UpstreamEndpoints` to be stored in a single list.

Here's some blog posts that go over the ECS stuff, in relation to what frameworks are available in Rust.

- https://csherratt.github.io/blog/posts/specs-and-legion/
- https://bevy-cheatbook.github.io/programming/ecs-intro.html